### PR TITLE
[BACKUP] Added support for Customer Managed Keys(CMK) for Recovery Services Vault

### PR DESCRIPTION
**Pull request by Vidhi-2510**
**Description**

The purpose of this PR was to introduce Customer Managed keys feature in Azure CLI. The following commandlets were created:

1. vault update – It is for setting the managed identities for Recovery services vaults. It has following parameters:
    -  identity_type {userassigned/systemassigned/none}
    -  identity_id {identityId}
    -  remove_system_assigned
    -  remove_user_assigned
    -  resource_group
    -  vault_name

2. vault encryption update – It is to update the encryption settings of Recovery services vault. It has following parameters:
    - encryption_key_id
    - identity_id
    - use_systemassigned_identity
    - infrastructure_encryption
    - resource_group
    - vault_name

3. vault encryption show – It is to show details of encryption properties of Recovery services vault. It has following parameters:
    - resource_group
    - vault_name

4. Restore-disk – Added disk-encryption-set-id parameter in restore_disks command for encrypting restored data with it.

Apart from the code for above features, the PR also contains the recordings of the required tests

**Testing Guide**

Identity commands

Addition of Identities
- az backup vault update --resource-group {rg} --vault-name {vault} --identity-type systemassigned
- az backup vault update --resource-group {rg} --vault-name {vault} --identity-type userassigned --identity-id <list_of_identity_id[]>

Removal of Identities
- az backup vault update --resource-group {rg} --vault-name {vault} --remove-system-assigned
- az backup vault update --resource-group {rg} --vault-name {vault} --remove-user-assigned --identity-id <list_of_identity_id[]>

Clear all identities assigned
- az backup vault update --resource-group {rg} --vault-name {vault} --identity-type none

Encryption commands

Set encryption
- az backup vault encryption update --resource-group {rg} --vault-name {vault} --encryption-key-id {encryptionkeyId} [--identity-id {identityId}|--use-system-assigned] [--infrastructure-encryption {Enabled, Disabled}]

Get encryption
- az backup vault encryption show --resource-group {rg} --vault-name {vault}

Restore command
- az backup restore restore-disks --storage-account {storageaccount} --container-name {container_name} --item-name {item_name} --resource-group {rg} --rp-name {rp} --target-resource-group {target_rg} --vault-name {vault} --disk-encryption-set-id {des} --restore-to-staging-storage-account true

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
BACKUP

File: src/azure-cli/azure/cli/command_modules/backup/_help.py [hunk 1]
```diff
@@ -495,3 +495,29 @@
     text: az backup vault show --name MyRecoveryServicesVault --resource-group MyResourceGroup
     crafted: true
 """
+helps['backup vault update'] = """
+type: command
+short-summary: Updates an existing Recovery Services vault.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_help.py:500]: Please use the first person voice
```suggestion
short-summary: Update an existing Recovery Services vault.
```

File: src/azure-cli/azure/cli/command_modules/backup/_help.py [hunk 2]
```diff
@@ -495,3 +495,29 @@
...
+examples:
+  - name: Updates an existing Recovery services vault. (autogenerated)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_help.py:502]:
```suggestion
  - name: Update an existing Recovery services vault. (autogenerated)
```

File: src/azure-cli/azure/cli/command_modules/backup/_help.py [hunk 3]
```diff
@@ -495,3 +495,29 @@
...
+    text: az backup vault update --identity-type None --resource-group MyResourceGroup --vault-name MyVault
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_help.py:503]: May I ask can the users not pass in the parameter  `-identity-type` instead of passing None?

File: src/azure-cli/azure/cli/command_modules/backup/_help.py [hunk 4]
```diff
@@ -495,3 +495,29 @@
...
+helps['backup vault encryption'] = """
+type: group
+short-summary: Encryption details of a Recovery Services Vault.
+"""
+helps['backup vault encryption update'] = """
+type: command
+short-summary: Updates encryption properties of a Recovery Services Vault.
+examples:
+  - name: Updates encryption properties of a Recovery Services Vault.
+    text: az backup vault encryption update --encryption-key-id MyEncryptionKeyId --resource-group MyResourceGroup --vault-name MyVault
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_help.py:515]:
```suggestion
short-summary: Update encryption properties of a Recovery Services Vault.
examples:
  - name: Update encryption properties of a Recovery Services Vault.
    text: az backup vault encryption update --encryption-key-id MyEncryptionKeyId --resource-group MyResourceGroup --vault-name MyVault
```

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 5]
```diff
@@ -46,6 +47,7 @@
 target_tier_help = """ The destination/target tier to which a particular recovery point has to be moved."""
 tier_help = """ Provide 'tier' parameter to filter recovery points."""
 rehyd_priority_type_help = """The type of priority to be maintained while rehydrating a recovery point """
+infrastructure_encryption_type_help = """Use this parameter to Enable/Disable Infrastructure encryption. This must be set when configuring encryption of the vault for the first time. Once Enabled/Disabled, infrastructure encryption setting cannot be changed. Default value: Disabled. Allowed values: Enabled /Disabled"""
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:50]:
```suggestion
infrastructure_encryption_type_help = """Use this parameter to Enable/Disable infrastructure encryption. This must be set when configuring encryption of the vault for the first time. Once Enabled/Disabled, infrastructure encryption setting cannot be changed. Default value: Disabled. Allowed values: Enabled /Disabled"""
```

**Review by zhoxing-ms — CHANGES_REQUESTED** [review]
@Vidhi-2510 Could you please design managed identity related commands according to our unified guideline?
[guideline link](https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md)

File: src/azure-cli/azure/cli/command_modules/backup/_help.py (hunk 3 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_help.py:503]: @zhoxing-ms user can pass None to remove all the identities at once if he wish. He can also pass SystemAssigned or UserAssigned as --identity-type for addition of identities

Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_help.py:503]: If you wish we can disallow passing the 'None' value to the identity-type. In our case 'None' is a value user passes to remove all identities in a vault.

**Comment by hiaga** [comment]
> @Vidhi-2510 Could you please design managed identity related commands according to our unified guideline?
> [guideline link](https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md)

@zhoxing-ms went through the guideline - looks good. We need one clarity though, what happens when the user give below command  - 

**az <resource> identity remove  --system-assigned --user-assigned**

does this remove both system-assigned/user-assigned identities? or throws error?

File: src/azure-cli/azure/cli/command_modules/backup/_help.py (hunk 3 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_help.py:503]: Got it, thanks~

**Comment by zhoxing-ms** [comment]
@hiaga  
> az [resource] identity remove --system-assigned --user-assigned
does this remove both system-assigned/user-assigned identities? or throws error?

Yes,  use `--system-assigned` to remove system assigned identity and `--user-assigned` with space separated resource IDs to remove specified user assigned identities. If both parameters are passed in, the system-assigned and user-assigned identities can be removed at the same time~

> az [resource] identity assign --system-assigned --user-assigned [AzureResourceId1]
can both the identities be added together ?

Yes, exactly!

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 6]
```diff
@@ -92,6 +98,33 @@ def load_arguments(self, _):
...
 
+    # Identity
+    with self.argument_context('backup vault identity assign') as c:
+        c.argument('vault_name', vault_name_type)
+        c.argument('system_assigned', action='store_true', options_list=['--system-assigned'], help="Provide this flag to enable system assigned identity for Recovery Services Vault.")
+        c.argument('user_assigned', user_assigned_type)
+
+    with self.argument_context('backup vault identity remove') as c:
+        c.argument('vault_name', vault_name_type)
+        c.argument('system_assigned', action='store_true', options_list=['--system-assigned'], help="Provide this flag to enable system assigned identity for Recovery Services Vault.")
```
Comment (hiaga) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:109]: Provide this flag to **remove** system assigned identity for Recovery Services Vault

File: src/azure-cli/azure/cli/command_modules/backup/custom.py [hunk 7]
```diff
@@ -106,7 +106,7 @@ def create_vault(client, vault_name, resource_group_name, location, tags=None):
     vault_sku = Sku(name=SkuName.standard)
     vault_properties = VaultProperties()
     vault = Vault(location=location, sku=vault_sku, properties=vault_properties, tags=tags)
-    return client.create_or_update(resource_group_name, vault_name, vault)
+    return client.begin_create_or_update(resource_group_name, vault_name, vault)
```
Comment (hiaga) [inline src/azure-cli/azure/cli/command_modules/backup/custom.py:109]: why is this change needed ?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 8]
```diff
@@ -71,6 +72,11 @@
 target_tier_type = CLIArgumentType(help=target_tier_help, arg_type=get_enum_type(allowed_target_tier_type_chk_archivable), options_list=['--target-tier'])
 tier_type = CLIArgumentType(help=tier_help, arg_type=get_enum_type(allowed_tier_type), options_list=['--tier'])
 rehyd_priority_type = CLIArgumentType(help=rehyd_priority_type_help, arg_type=get_enum_type(allowed_rehyd_priority_type), options_list=['--rehydration-priority'])
+user_assigned_identity_id_type = CLIArgumentType(options_list=['--identity-id'], help="UserAssigned Identity Id to be used for CMK encryption, this will be applicable for encryption using userassigned identity")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: Could we refer to the recommendation in the [guideline.md#enable-managed-identity-during-resource-creation](https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md#enable-managed-identity-during-resource-creation), change the parameter name to `--mi-user-assigned`

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 9]
```diff
@@ -71,6 +72,11 @@
...
+encryption_key_id_type = CLIArgumentType(options_list=['--encryption-key-id'], help="The encryption key id you want to use for encryption")
+infrastructure_encryption_type = CLIArgumentType(options_list=['--infrastructure-encryption'], arg_type=get_enum_type(['Enabled', 'Disabled']), help=infrastructure_encryption_type_help)
+user_assigned_type = CLIArgumentType(nargs='+', options_list=['--user-assigned'], help="Space-separated list of userassigned identities.")
+user_assigned_remove_type = CLIArgumentType(nargs='*', options_list=['--user-assigned'], help="Space-separated list of userassigned identities.")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:79]: Can `user_assigned_type` be replaced by `user_assigned_remove_type`, so it's redundant?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 10]
```diff
@@ -92,6 +98,33 @@ def load_arguments(self, _):
         c.argument('soft_delete_feature_state', arg_type=get_enum_type(['Enable', 'Disable']), help='Set soft-delete feature state for a Recovery Services Vault.')
         c.argument('cross_region_restore_flag', arg_type=get_enum_type(['True', 'False']), help='Set cross-region-restore feature state for a Recovery Services Vault. Default: False.')
 
+    # Identity
+    with self.argument_context('backup vault identity assign') as c:
+        c.argument('vault_name', vault_name_type)
+        c.argument('system_assigned', action='store_true', options_list=['--system-assigned'], help="Provide this flag to enable system assigned identity for Recovery Services Vault.")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:104]: Could we also define a CLIArgumentType for `system_assigned` just like `user_assigned_type`？

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 11]
```diff
@@ -92,6 +98,33 @@ def load_arguments(self, _):
...
+
+    # Encryption
+    with self.argument_context('backup vault encryption') as c:
+        c.argument('vault_name', vault_name_type)
+
+    with self.argument_context('backup vault encryption update') as c:
+        c.argument('encryption_key_id', encryption_key_id_type)
+        c.argument('infrastructure_encryption', infrastructure_encryption_type)
+        c.argument('identity_id', user_assigned_identity_id_type)
+        c.argument('use_systemassigned_identity', action='store_true', options_list=['--use-system-assigned'], help="Provide this flag to use system assigned identity for encryption.")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:123]: Could we refer to the recommendation in the [guideline.md#enable-managed-identity-during-resource-creation](https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md#enable-managed-identity-during-resource-creation), change the parameter name to `--mi-system-assigned`

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 8 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: @zhoxing-ms : we don't have creation time experience for managed identities. This parameter is used for Customer managed key encryption properties for Vault

File: src/azure-cli/azure/cli/command_modules/backup/tests/latest/test_backup_commands.py [hunk 12]
```diff
@@ -749,3 +749,455 @@ def test_backup_archive (self, resource_group, vault_name, vm_name, storage_acco
             self.check("tierType", '{rp1_tier}'),
             self.check("resourceGroup", '{rg}')
         ])
+    # @ResourceGroupPreparer()
+    # @VaultPreparer()
+    # def test_backup_identity(self, resource_group, vault_name):
+    #     self.kwargs.update({
+    #         'vault': vault_name,
+    #         'rg': resource_group,
+    #     })
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/tests/latest/test_backup_commands.py:758]: If these tests are no longer needed, why not delete them directly?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 9 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:79]: @zhoxing-ms: we want to keep nargs different - is there a better way to do this ?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 8 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: Yes, because this parameter is the uniform name for adding user-assigned identities when creating and updating resources.

File: src/azure-cli/azure/cli/command_modules/backup/tests/latest/test_backup_commands.py (hunk 12 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/tests/latest/test_backup_commands.py:758]: sure, will remove the commented tests

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 8 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: Actually, this parameter `--mi-user-assigned` is the uniform name for adding user-assigned identities when creating and updating resources.
Do you mean that the parameter `--identity-id` has been used in other similar places?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 11 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:123]: @zhoxing-ms : we don't have creation time experience for managed identities in the vault. This parameter is used for Customer managed key encryption properties for Vault

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 9 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:79]: Could we not add `nargs` to `CLIArgumentType`, but declare it separately for different commands?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 8 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: backup vault encryption update --encryption-key-id {key1_id} --identity-id {identity1_id} -g {rg} -v {vault1}

we are using this parameter for the encryption group

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:75]: OK, I see. It is still recommended to use `--mi-user-assigned` in line with the guideline

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 11 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:123]: Same as above

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 13]
```diff
@@ -71,6 +72,14 @@
 target_tier_type = CLIArgumentType(help=target_tier_help, arg_type=get_enum_type(allowed_target_tier_type_chk_archivable), options_list=['--target-tier'])
 tier_type = CLIArgumentType(help=tier_help, arg_type=get_enum_type(allowed_tier_type), options_list=['--tier'])
 rehyd_priority_type = CLIArgumentType(help=rehyd_priority_type_help, arg_type=get_enum_type(allowed_rehyd_priority_type), options_list=['--rehydration-priority'])
+mi_user_assigned_type = CLIArgumentType(options_list=['--mi-user-assigned'], help="UserAssigned Identity Id to be used for CMK encryption, this will be applicable for encryption using userassigned identity")
+mi_system_assigned_type = CLIArgumentType(action='store_true', options_list=['--mi-system-assigned'], help="Provide this flag to use system assigned identity for encryption.")
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:76]: Could you please add examples for these two parameters?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 9 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py]: Such as:
```Python
with self.argument_context('backup vault identity assign') as c:
    c.argument('user_assigned', nargs='+', user_assigned_type)

with self.argument_context('backup vault identity remove') as c:
    c.argument('user_assigned',nargs='*',  user_assigned_type)
```

Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py]: as discussed - keeping it as it is because we no longer have same  user_assigned_type help for both params

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 13 above)
Comment (hiaga) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:76]: done

**Review by zhoxing-ms — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]
